### ISSUE TITLE
Always connects to the master due to the wrong detection of LudicrousDB->is_write_query()

### ISSUE THREAD

**Issue by ooxif** [stuttter/ludicrousdb#94]
WordPress always executes a query `SELECT option_name, option_value FROM infy_options WHERE autoload = 'yes'`, and LudicrousDB->is_write_query() returns a false positive result, true.

Because of that, WordPress always connects to the master, which results in a connection flood to the master.

Related: #55, #70

**Comment by takotakot** [comment]
@ooxif Sorry for my misconstruction. Does the change from

```
preg_match( '/(?:ALTER|CREATE|ANALYZE|CHECK|OPTIMIZE|REPAIR|CALL|DELETE|DROP|INSERT|LOAD|REPLACE|UPDATE|SET|RENAME)\s/i', $q )
```

to

```
preg_match( '/(?:^|\s)(?:ALTER|CREATE|ANALYZE|CHECK|OPTIMIZE|REPAIR|CALL|DELETE|DROP|INSERT|LOAD|REPLACE|UPDATE|SET|RENAME)\s/i', $q )
```

help? Or any ideas?

**Comment by ooxif** [comment]
@takotakot LGTM, thanks!

**Comment by ooxif** [comment]
@takotakot 
https://github.com/sysown/proxysql/wiki/ProxySQL-Read-Write-Split-(HOWTO)#basic-readwrite-split-using-regex
ProxySQL does a special exception only for `SELECT ... FOR UPDATE` statement.

```php
return ! preg_match( '/^\s*SELECT\s.*\sFOR UPDATE\*$/', $q ) && ! preg_match( '/^(?:SELECT|SHOW|DESCRIBE|DESC|EXPLAIN)\s/i', $q );
```

migit help, IMO

**Comment by takotakot** [comment]
@ooxif We can write:

```
SELECT 1 FROM table1;UPDATE table1 SET col=1;
```

. It shows that "begin with `SELECT`" may be bad idea...

Added ';'
```
preg_match( '/(?:^|\s|;)(?:ALTER|CREATE|ANALYZE|CHECK|OPTIMIZE|REPAIR|CALL|DELETE|
DROP|INSERT|LOAD|REPLACE|UPDATE|SET|RENAME)\s/i', $q )
```

**Comment by ooxif** [comment]
@takotakot 
- Neither [`mysql_query` nor `mysqli_query` (used in LudicrousDB->_do_query)](https://github.com/stuttter/ludicrousdb/blob/5.0.0/ludicrousdb/includes/class-ludicrousdb.php#L1480) supports executing multiple queries at once. `$query` is expected not to have semicolons (statement separators) because such $query is invalid.
- "begin with XXX" regexp patterns such as `preg_match( '/^\s*SELECT\s+....` and `preg_match( '^\s(create|...` are widely used in [class-ludicrousdb.php](https://github.com/stuttter/ludicrousdb/blob/5.0.0/ludicrousdb/includes/class-ludicrousdb.php).

I think we can omit that case.

**Comment by takotakot** [comment]
I made two versions: #95, #96 .

Maintainers:
Please choose the better one or comment here.

**Comment by JJJ** [comment]
#95 seems viable, but needs testing to avoid regressions.

**Comment by ooxif** [comment]
I commented another problem about is_write_query() at https://github.com/stuttter/ludicrousdb/issues/55#issuecomment-574520356
FYI, please take a look at the comment.

**Comment by JJJ** [comment]
@ooxif I've merged #100 into master which should fix the trailing space issue. Give it a whirl?

**Comment by ooxif** [comment]
@JJJ Thank you for your response!
But I have used [my own patch](https://github.com/ooxif/ludicrousdb/compare/master...ooxif:fix/is-write-query) in production for months and it's going well.
So I'll go with this.
#100 LGTM, though.

**Comment by archon810** [comment]
Is 5.1 getting released soon?

**Comment by JJJ** [comment]
> Is 5.1 getting released soon?

It would be ideal to have #95 resolved for 5.1.

I will close this issue since the OP is OK with the fix from #100.
